File: src/solutions/2025/03/python/main.py
import sys
import os
import typing

sys.path.append(os.getcwd())
from src.langs.python.run import run


def main() -> None:
   options = {'has_alternate': False, 'has_io': False}
   run(sys.argv, part_one, part_two, options)


def solve(input: str, is_test: bool, p2: bool) -> str:
   res = 0
   for line in input.split('\n'):
      start = 0
      max = 12 if p2 else 2
      b = line.encode('utf-8')
      for digit in range(max):
         marked = 0
         n = 0
         t = max - 1 - digit
         l = len(line) - t
         for it in range(start, l):
            parsed = b[it] - int.from_bytes(b'0')
            if n < parsed:
               marked = it
               n = parsed
         start = marked + 1
         res += n * (10**t)

   return str(res)


def part_one(input: str, is_test: bool) -> str:
   return solve(input, is_test, False)


def part_two(input: str, is_test: bool) -> str:
   return solve(input, is_test, True)


if __name__ == '__main__':
   main()

File: src/solutions/2025/03/ts/main.ts
import { argv } from "node:process";
import type { SolutionOptions } from "src/options.ts";
import { run } from "src/run.ts";

const options: SolutionOptions = {
   hasAlternate: false,
   hasIo: false,
};

function solve(input: string, isTest: boolean, p2: boolean): string {
   const textEncoder = new TextEncoder();
   return input
      .split("\n")
      .map((line) => textEncoder.encode(line))
      .reduce((acc, line) => {
         let res = 0;
         let start = 0;
         const max = p2 ? 12 : 2;
         for (let digit = 0; digit < max; digit++) {
            let marked = 0;
            let n = 0;
            const t = max - 1 - digit;
            const l = line.length - t;
            for (let it = start; it < l; it++) {
               const parsed = line[it] - 0x30;
               if (n < parsed) {
                  marked = it;
                  n = parsed;
               }
            }
            start = marked + 1;
            res += n * Math.pow(10, t);
         }
         return acc + res;
      }, 0)
      .toString();
}

function part1(input: string, isTest: boolean): string {
   return solve(input, isTest, false);
}

function part2(input: string, isTest: boolean): string {
   return solve(input, isTest, true);
}

if (import.meta.main) {
   run(argv, part1, part2, options);
}
